server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -273,11 +396,24 @@ public void setFielddataMinSegmentSize(int fielddataMinSegmentSize) {
             this.fielddataMinSegmentSize = fielddataMinSegmentSize;
         }
 
+        void setPrefixFieldType(PrefixFieldType prefixFieldType) {
+            checkIfFrozen();
+            this.prefixFieldType = prefixFieldType;
+        }
+
         @Override
         public String typeName() {
             return CONTENT_TYPE;
         }
 
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (prefixFieldType == null || prefixFieldType.accept(value.length()) == false) {
+                return super.prefixQuery(value, method, context);
+            }
+            return prefixFieldType.termQuery(value, context);
```

Comment:
should we use constant score as well if the rewrite method is null or constant_score ?